gslib/command.py

```diff
@@ -725,27 +729,47 @@ def RaiseInvalidArgumentException(self):
                (self.command_spec.usage_synopsis, self.command_name))
     raise CommandException(message)
 
-  def ParseSubOpts(self, check_args=False):
+  def ParseSubOpts(self,
+                   check_args=False,
+                   args=None,
+                   update_sub_opts_and_args=True):
     """Parses sub-opt args.
 
     Args:
       check_args: True to have CheckArguments() called after parsing.
-
-    Populates:
-      (self.sub_opts, self.args) from parsing.
+      args: List of args. If None, self.args will be used.
+      update_sub_opts_and_args: True if self.sub_opts and self.args should
+        be updated with the values returned after parsing. Else return a 
+        tuple of sub_opts, args returned by getopt.getopt. This is done
+        to allow this method to be called from GetGcloudStorageArgs in which
+        case we do not want to update self.sub_opts and self.args.
 
     Raises: RaiseInvalidArgumentException if invalid args specified.
     """
+    if args is None:
+      unparsed_args = self.args
+    else:
+      unparsed_args = args
     try:
-      self.sub_opts, self.args = getopt.getopt(
-          self.args, self.command_spec.supported_sub_args,
+      parsed_sub_opts, parsed_args = getopt.getopt(
+          unparsed_args, self.command_spec.supported_sub_args,
           self.command_spec.supported_private_args or [])
```

Comment:
I am afraid of `or []` for the same reason you shouldn't do `def f(x=[])`.
I'm concerned it creates one object that will have old data if this section of code runs again.
Did you experiment to see if it would persist data?

Reply:
I think you are talking about the default argument case `def f(x=[])`. This is not same as default argument because here we are deciding the argument while calling the method. So it should be fine and there is no risk.